topics/data-science/tutorials/python-argparse/tutorial.md

```diff
@@ -0,0 +1,248 @@
+> {: .solution}
+{: .question}
+
+## Simple tasks
+
+Let's sum up all of the numbers passed on the command line. We'll do this by hand, and then we'll replace it with `argparse` to see how much effort that saves us.
+
+> ### {% icon hands_on %} Hands-on
+> Update your script to sum up every number passed to it on the command line.
+>
+> It should handle:
+> - 1 or more numbers
+> - nothing (and maybe print out a message?)
+> - invalid values (print out an error message that the value couldn't be processed.)
+>
+> Hints:
+> - Skip the program name
+> - Use `try` and `except` to try converting the string to a number.
+> > ### {% icon question %} Question
+> > How does your updated script look?
+> > > ### {% icon solution %} Solution
+> > >
+> > > ```python
+> > > import sys
+> > >
+> > > result = 0
+> > >
+> > > if len(sys.argv) == 1:
+> > >     print("no arguments were supplied")
+> > > else:
+> > >     for arg in sys.argv[1:]:
+> > >         try:
+> > >             result += float(arg)
+> > >         except:
+> > >             print(f"Could not parse {arg}")
+> > >
+> > >     print(result)
+> > > ```
+> > {: .solution}
+> {: .question}
+{: .hands_on}
+
+## Argparse
+
+Argparse saves us a lot of work, because it can handle a number of things for us!
+
+- Ensures that the correct number of arguments are provided (and provide a nice error message otherwise)
+- Ensure that the correct types of arguments are provided (no strings for a number field)
+- Provide a help message describing your program
+
+Argparse is used as follows. First we need to import it
+
+```python
+import argparse
+```
+
+And then we can define a 'parser' which will parse our command line. Additionally we can provide a description field which tells people what our tool does:
+
+```python
+parser = argparse.ArgumentParser(description='Process some integers.')
+```
+
+And finally we can define some arguments that are available. Just like we have arguments to functions, we have arguments to command lines. These come in two flavours:
+
+- required (without a `--`)
+- optional "flags" (prefixed with `--`)
+
+Here we have an argument named 'integers', which validates that all input values are of the type `int`. `nargs` is the number of arguments, `+` means '1 or more'. And we have some help text as well:
+
+```python
+parser.add_argument('integers', type=int, nargs='+',
+                    help='an integer for the accumulator')
+```
+
+We can also define an optional flag, here it's called `--sum`. Here it goes to a destination named 'accumulate', the name we'll use to access the value of this argument. It has an action of 'store_const' which just tracks if the flag was supplied or not.
+
+The `const` attribute is set to `sum`, which is actually the function `sum()`, this is what the value will be if we run the command with `--sum`. Otherwise it will `default` to the function `max()`. We again have some help text to tell us how it behaves
+
+```python
```

Comment:
can we perhaps see this argument part in action?

little example with a script and how it behaves?